tests - add test for ResourceWarning: unclosed transport <asyncio.sslproto._SSLProtocolTransport object …

commonism opened the pull request:
## What do these changes do?

Add a test for 
`ResourceWarning: unclosed transport <asyncio.sslproto._SSLProtocolTransport object at …`
when using aiohttp.ClientSession with TLS

## Are there changes in behavior for the user?

No

## Related issue number

#4703
#4713

commonism committed:
tests - add test for ResourceWarning: unclosed transport <asyncio.sslproto._SSLProtocolTransport object …

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+tracemalloc.start(40)
+
+
+log = logging.getLogger()
+log.setLevel(logging.DEBUG)
+logging.basicConfig(level=logging.WARNING)
```
It's pretty uncommon to have random logs in tests. Also, such a setup is going to influence the whole runtime which is bad.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+            except Exception as e:
+                log.exception(e)
+            finally:
+                await asyncio.sleep(5)
```
Making each test invocation 5 seconds slower doesn't sound like a great idea.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+    return _setUp, _tearDown
+
+
+class TestUnclosedTransport(unittest.IsolatedAsyncioTestCase):
```
Inheriting from `unittest.TestCase` and similar prevents pytest from using its features and must be avoided.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+
+    @routes.get('/', name='root')
+    async def handle_root(self, request):
+        import random
+        off = random.randint(7,17)
+        await asyncio.sleep(off/10)
+        return web.Response(text=f'{off}', status=200)
```
What is this supposed to do exactly?

commonism replied:
Force a connection_lost.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+        runner = web.AppRunner(app)
+        self.runner = runner
+        self.server_ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+        self.server_ctx.load_cert_chain('data/service-certificates.pem', 'data/service-key.pem')
```
Referencing non-existent files isn't going to work. Also, hardcoding pre-generated files isn't great either. In this test suite, there's good examples of making temporary TLS certificates: we use `trustme` lib for this but you don't need to use it directly because there already a set of pytest fixtures that are ready-to-use — use them instead. Usage example: https://github.com/aio-libs/aiohttp/blob/32833c3/tests/test_client_functional.py#L295-L297.

webknjaz commented:
@commonism Git setup on your computer looks broken: none of the submitted commits list your GitHub account as an author. You should probably configure Git locally or verify the email that is embedded in those commits in your GitHub account.

webknjaz reviewed (changes requested):
Thanks for the effort! Unfortunately, this doesn't look ready, nor is compliant with our test writing practices.

Also, there's great prior art on the usage of pytest fixtures including proper TLS setup — https://github.com/aio-libs/aiohttp/pull/4713/files — why haven't you start there? It also uses proper loop isolation, and pytest-based warnings tracking.

Besides, adding a custom server/client implementations doesn't seem related to the problem you're attempting to reproduce — it only makes the test module messy and distracts from what's actually being tested.

asvetlov reviewed (commented):
@commonism I respect your striving to point on the very annoying issue.
It will be never fixed on aiohttp 3.x branch, sorry.
aiohttp 4.0 has partially done work for getting rid of the warning. The missing part is #5102

P.S.
aiohttp uses `pytest` with `pytest-aiohttp` plugin, not the standard `unittest`.
This is the deliberate choice. We don't accept tests that are `unittest` based.

P.P.S.
I'm aware of `unittest.IsolatedAsyncioTestCase` because I had added it to the CPython standard library.
It exists because CPython itself cannot depend on `pytest` project for running self-testing.
For third-party libs I always prefer pytest-style.

commonism commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,156 @@
+tracemalloc.start(40)
+
+
+log = logging.getLogger()
+log.setLevel(logging.DEBUG)
+logging.basicConfig(level=logging.WARNING)
```
Will you accept the test if the defects outlined get resolved?

asvetlov replied:
Do you mean `pytest` usage also?

commonism replied:
Yes, including everything mentioned so far.

commonism reviewed (commented):
I addressed the following issues

 - pytest, not a unittest
 - no logging/print
 - no tracemalloc
 - reduce time required to 5s (was 30 seconds actually)
 - using the internal trustme CA

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    idle 2 seconds so service is restarted meanwhile, forcing connection_lost
+    """
+    await asyncio.sleep(2)
+    return web.Response(text='', status=200)
```
status is redundant
```suggestion
    return web.Response(text='')
```

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+        finally:
+            await asyncio.sleep(0.1)
+
+        assert [warnings.formatwarning(w.message, w.category, w.filename, w.lineno, w.line)
+                for w in filter(_ssl_resource_warnings, recwarn)] == []
```
It seems the check you're after is
```suggestion
        assert not len(filter(_ssl_resource_warnings, recwarn))
```
I don't see any value in converting warning objects that you are only going to count and throw away.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    return None
+
+async def _run(loop, recwarn, ssl_ctx, client_ssl_ctx):
+    global been_restarted
```
Globals must be avoided per the same concerns as before: mutating global state makes tests unisolated.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+            await asyncio.sleep(0.5)
+
+def _ssl_resource_warnings(w):
+    if w.category == ResourceWarning and w.filename.endswith('sslproto.py') and \
+        w.message.args[0].startswith('unclosed transport <asyncio.sslproto._SSLProtocolTransport object at'):
+        return w
+    return None
```
Aren't filter predicates [supposed to return bool](https://docs.python.org/3/library/functions.html#filter)?

webknjaz replied:
So the proper implementation would be
```suggestion
    unclosed_transport_msg = 'unclosed transport <asyncio.sslproto._SSLProtocolTransport object at'
    return (
        w.category == ResourceWarning
        and w.filename.endswith('sslproto.py')
        and w.message.args[0].startswith(unclosed_transport_msg)
    )
```

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+async def _prepare(ssl_ctx, client_ssl_ctx):
+    app = web.Application()
+    app.router.add_get('/', handle_root)
+    runner = web.AppRunner(app)
```
It's better not to reimplement the server initialization. Just use our native [`aiohttp_server` fixture](https://github.com/aio-libs/aiohttp/blob/f8df85d/aiohttp/pytest_plugin.py#L261). It at least will get you a server with an app runner set up and you can probably extend it with a tcp site on top.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    await _close(app, runner, session)
+
+
+def test_run(recwarn, ssl_ctx, client_ssl_ctx):
```
This test must have some meaningful name telling the reader that it is actually for.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+                await s.stop()
+                await s.start()
+            been_restarted += 1
+        except Exception as e:
+            print(e)
```
What is suppressing all the exceptions for?

commonism reviewed (commented):
addressing the issues.

webknjaz commented:
@commonism we also need the CI to actually execute this test. Red linting job prevents that from happening.

webknjaz commented:
@commonism P.S. your local Git setup is still broken

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    app['dq'] = dq
+    app.router.add_get('/', handle_root)
+    server = await aiohttp_server(app, port=12345, ssl=ssl_ctx)
+    asyncio.ensure_future(_restart(server.runner, ssl_ctx, cq))
```
I don't get why you need to schedule a task on the loop and not wait for it. You're basically creating a synthetic race condition that will likely make the testing indeterministic.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    app['cq'] = cq
+    app['dq'] = dq
+    app.router.add_get('/', handle_root)
+    server = await aiohttp_server(app, port=12345, ssl=ssl_ctx)
```
It's better to use an ephemeral port 0. You can then get a client to whatever the actual port ends up being via [`aiohttp_client` fixture](https://github.com/aio-libs/aiohttp/blob/f8df85d/aiohttp/pytest_plugin.py#L334) and you don't need to care about managing the client session manually.

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+    while True:
+        for s in list(runner.sites):
+            await s.stop()
+            s = web.TCPSite(runner, '127.0.0.1', port=12345, ssl_context=ssl_ctx)
```
What is the point of creating a new site object on the same port?

commonism replied:
closing the site and re-opening the site to force a client side connection_lost

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+
+        assert not len(list(filter(_ssl_resource_warnings, recwarn)))
+
+    await _close(server, session)
```
Since this is not executed in a `finally:` block, it won't get executed in case of assertion errors potentially messing up external tests or anything that hardcodes its port to 12345.

commonism replied:
using port 0 now

webknjaz commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,83 @@
+
+    while dq.qsize() <= 2:
+        try:
+            await session.get('https://localhost:12345/')
```
Don't hardcode hostname/port. Just use a built-in test client from the fixture I've mentioned before.

commonism commented:
> @commonism we also need the CI to actually execute this test. Red linting job prevents that from happening.

The Red Linter is gone, now the tests fail after 2 minutes - timeout?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aio-libs/aiohttp?pullRequest=5146) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/aio-libs/aiohttp?pullRequest=5146) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: common<br/>:x: commonism<br/><hr/>**common** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aio-libs/aiohttp?pullRequest=5146) it.</sub>

commonism commented:
Rebased, addressed the authorship issues to have CLAassistant compliance, force pushed.

"Timeline protection" is missing, following the Details, the link https://camo.githubusercontent.com/81006649ffdf2304b123fbe39fca080c91f9aef3e8e4d33a0958e88c04d41a16/68747470733a2f2f7468656576656e746368726f6e69636c652e636f6d2f77702d636f6e74656e742f75706c6f6164732f323031342f31302f7661746c6962372e6a7067 is "Non-Image content-type returned"

Any advice?

commonism commented:
This is still an issue.
How to proceed to get this test merged.
Maybe as pytest.mark.fail?

Dreamsorcerer commented:
We've got some stricter type checking these days. If you could fix the mypy errors and add an xfail, then I'll make sure this gets merged.

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,118 @@
+import asyncio
+
+import aiohttp
```
Could probably just import the 2 things we need from the module here.

commonism committed:
mypy - type annotations

commonism committed:
xfail

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,144 @@
+    client_ssl_ctx: Any,
+    cq: asyncio.Queue[int],
+    dq: asyncio.Queue[int],
+) -> Tuple[TestServer, TestClient]:
```
I know many of our tests are still labelled with `Any`, but I'm in the process of cleaning it up. Could you add in the correct annotations here? Let me know if there's anything you can't find the type of.

Off the top of my head, TestClient/TestServer are in aiohttp.test_utils, ssl_ctx is `ssl.SSLContext`, and any other fixtures can be checked in the source code (they should have the return type annotated correctly).

commonism replied:
try regexxer for the cleanup.

Current
```
AiohttpClient = Callable[[Union[Application, BaseTestServer]], Awaitable[TestClient]]
AiohttpServer = Callable[[Application], Awaitable[TestServer]]
```
do not have the kwargs parameters for port & ssl_ctx, which is why I have to declare them as Any in one location.

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,144 @@
+    await _close(server, session)
+
+
+@pytest.mark.xfail
```
```suggestion
@pytest.mark.xfail(reason="Depends on #5102")
```

commonism committed:
mypy - typing beyond Any

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,151 @@
+    server, session = await _prepare(
+        aiohttp_server, aiohttp_client, ssl_ctx, client_ssl_ctx, cq, dq
+    )
+    if server.runner is None or session.port is None:
+        # mypy | None required
+        return
```
```suggestion
    assert server.runner is not None and session.port is not None
```

Dreamsorcerer committed:
Update tests/test_ssl_transport_close.py

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,149 @@
+            jobs.append(_restart(server.runner, ssl_ctx, session.port, cq, dq))
+            await asyncio.gather(*jobs)
+        except ServerDisconnectedError:
+            """
+            restarting the service will cause the client connections to fail
+            this is expected and no failure
+            """
```
```suggestion
            # Restarting the service will cause the client connections to fail
            # this is expected and not a failure.
```

Dreamsorcerer committed:
Update tests/test_ssl_transport_close.py

Dreamsorcerer committed:
Update test_ssl_transport_close.py

Dreamsorcerer commented:
```
 E               Traceback (most recent call last):
E                 File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/warnings.py", line 537, in _warn_unawaited_coroutine
E                   warn(msg, category=RuntimeWarning, stacklevel=2, source=coro)
E               RuntimeWarning: coroutine 'aiohttp_server.<locals>.finalize' was never awaited
```

Because it's a cleanup thing, the error appears in the next test, but it's caused by this test.

Dreamsorcerer commented:
Seemingly only on Python 3.11+ for some reason though.

commonism committed:
fix - coroutine 'aiohttp_server.<locals>.finalize' was never awaited

commonism commented:
Should be fixed

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,148 @@
+async def _close(server: TestServer, session: TestClient) -> None:
+    await server.close()
+    await session.close()
+    await server
```
I don't think that fixes the issue, but somehow masks it.
```
TypeError: object TestServer can't be used in 'await' expression
```

Dreamsorcerer replied:
I'm also seeing:
```
Exception ignored in: <coroutine object aiohttp_client.<locals>.finalize at 0x7fe61a202340>
Traceback (most recent call last):
  File "<string>", line 1, in <lambda>
KeyError: '__import__'
Exception ignored in: <coroutine object aiohttp_client.<locals>.finalize at 0x7fe61a202340>
Traceback (most recent call last):
  File "<string>", line 1, in <lambda>
KeyError: '__import__'
```
Which must be related to the problem.

Dreamsorcerer replied:
Think I've got a solution. The server/client already auto-close in the fixture, but the test was mixing up loops.

Dreamsorcerer replied:
So, actually, the test passes on 3.11+, which is why the warning didn't happen on the other tests.

Dreamsorcerer committed:
Fix loop

Dreamsorcerer committed:
Put xfail back in

Dreamsorcerer committed:
Update test_ssl_transport_close.py

Dreamsorcerer commented:
Hmm, there's a TimeoutError in 3.12 though...
Not sure what's happening there. It also passes on pypy though.

Dreamsorcerer committed:
Update test_ssl_transport_close.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dreamsorcerer committed:
Update test_ssl_transport_close.py

Dreamsorcerer commented:
Don't suppose you have an ideas about the TimeoutError on 3.12?
Once we figure that out we can merge.

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,141 @@
+            runner, "127.0.0.1", port=port, ssl_context=ssl_ctx, shutdown_timeout=1.0
+        )
+        await s.start()
+        break
```
Just noticed this one in the codecov.
The break means this never iterates, so there's no point of a loop here.

Dreamsorcerer replied:
If the intention is just to use any site from the set, then try `site = next(iter(runner.sites))`.

commonism replied:
We only have 1 site.

`site = next(iter(runner.sites))` would be limited to 1 site as well?

Dreamsorcerer replied:
Yes, my point is that, if we are not iterating, then we don't need a loop. Which will also clear up coverage, which highlights the for loop as a partial, because it only runs once.

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,141 @@
+    """restart service to force connection_lost"""
+    await dq.get()
+    dq.task_done()
+    for s in list(runner.sites):
+        await s.stop()
+        await cq.put(0)
+        s = web.TCPSite(
+            runner, "127.0.0.1", port=port, ssl_context=ssl_ctx, shutdown_timeout=1.0
+        )
+        await s.start()
+        break
```
```suggestion
    site = next(iter(runner.sites))
    await site.stop()
    await cq.put(0)
    site = web.TCPSite(
        runner, "127.0.0.1", port=port, ssl_context=ssl_ctx, shutdown_timeout=1.0
    )
    await site.start()
```

commonism committed:
Update tests/test_ssl_transport_close.py

Co-authored-by: Sam Bull <git@sambull.org>

commonism commented:
> Hmm, there's a TimeoutError in 3.12 though... Not sure what's happening there. It also passes on pypy.

Actually the timer got canceld?

```
self = <aiohttp.helpers.TimerContext object at 0x7ff31767dac0>
exc_type = <class 'asyncio.exceptions.CancelledError'>
exc_val = CancelledError(), exc_tb = <traceback object at 0x7ff3192e1d00>

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[TracebackType],
    ) -> Optional[bool]:
        if self._tasks:
            self._tasks.pop()  # type: ignore[unused-awaitable]
    
        if exc_type is asyncio.CancelledError and self._cancelled:
>           raise asyncio.TimeoutError from None
E           TimeoutError

exc_tb     = <traceback object at 0x7ff3192e1d00>
exc_type   = <class 'asyncio.exceptions.CancelledError'>
exc_val    = CancelledError()
self       = <aiohttp.helpers.TimerContext object at 0x7ff31767dac0>

aiohttp/helpers.py:723: TimeoutError
```

https://github.com/aio-libs/aiohttp/actions/runs/5978392347/job/16220898181

Dreamsorcerer commented:
> Actually the timer got canceld?

Hmm, we switched to asyncio.timeout() recently. So, I've merged master and will see if that helps.

commonism commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,140 @@
+            # this is expected and not a failure.
+            pass
+        finally:
+            await asyncio.sleep(0.1)
```
I guess it's the 300s ClientTimeout

        real_timeout = ClientTimeout(total=300, connect=None, sock_read=None, sock_connect=None, ceil_threshold=5)

300.26s call     tests/test_ssl_transport_close.py::test_unclosed_transport_asyncio_sslproto_SSLProtocolTransport[pyloop]


increasing the sleep to something larger to have to server restart in time?

Dreamsorcerer replied:
I'm not seeing how the sleep relates to it, but we can try it if you think that's the problem.

commonism commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,140 @@
+    await dq.get()
+    dq.task_done()
+    site = next(iter(runner.sites))
+    await site.stop()
```
In 3.12 it hangs here - you do not get 

> DEBUG    aiohttp.server:web_protocol.py:434 Ignored premature client disconnection

as you get in 3.11

Dreamsorcerer replied:
I don't see anything suspicious in there, unless the asyncio server is not closing:
https://github.com/commonism/aiohttp/blob/ebe2a48c282f2484ee1766af5c4d52ff3431b43a/aiohttp/web_runner.py#L83

Everything else looks like it has a 60s timeout, so would fail before the 5 min timeout we're seeing.

Dreamsorcerer replied:
If it is the asyncio server not closing, then we should get a bug filed for the regression.

Dreamsorcerer committed:
Merge branch 'master' into unclosedssltransportwarning

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,140 @@
+    await site.stop()
+    await cq.put(0)
+    site = web.TCPSite(
+        runner, "127.0.0.1", port=port, ssl_context=ssl_ctx, shutdown_timeout=1.0
```
```suggestion
        runner, "127.0.0.1", port=port, ssl_context=ssl_ctx
```

Dreamsorcerer committed:
Update tests/test_ssl_transport_close.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dreamsorcerer commented on tests/test_ssl_transport_close.py:
```diff
@@ -0,0 +1,138 @@
+        ), "unclosed transport"
+
+
+@pytest.mark.xfail(
+    sys.version_info < (3, 11) and platform.python_implementation() != "PyPy",
+    reason="Working on 3.11+ and Pypy.",
+)
```
```suggestion
```

Dreamsorcerer committed:
Update tests/test_ssl_transport_close.py

Dreamsorcerer commented:
This is actually working now. But, I don't know if we want to merge such a complex test after realising that these warnings were already visible in many tests (just ignored in the pytest config), so we already have coverage of this.
